### PR TITLE
refactor: implement getCode in Notice base class

### PR DESCRIPTION
**Pull request by lionel-nj**
closes #783 

**Summary:**

This PR refactors `Notice.getCode`: getCode is implemented in Notice base class.

**Expected behavior:** 

The notice code derives from its class name and formatted_as_this. 
Following 

> Step 1. Implement Notice.getCode() method. **It should strip the "notice" substring but not strip "error" since we will carefully remove that suffix in the further steps.**

https://github.com/MobilityData/gtfs-validator/pull/783#issuecomment-790123561

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- ~[] Include screenshot(s) showing how this pull request works and fixes the issue(s)~

### CODE REVIEW

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 1]
```diff
@@ -16,13 +16,15 @@
...
 
+import com.google.common.base.CaseFormat;
 import com.google.common.base.Joiner;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Objects;
 
 /** Base class for all notices produced by GTFS validator. */
 public abstract class Notice {
+  private static String NOTICE_SUBSTRING = "notice";
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:27]: This variable should also be `final`. I would also rename it to `NOTICE_SUFFIX`.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 2]
```diff
@@ -40,11 +42,18 @@ public SeverityLevel getSeverityLevel() {
   }
 
   /**
-   * Returns a descriptive type-specific name for this notice.
+   * Returns a descriptive type-specific name for this notice based on the class simple name.
    *
    * @return notice code, e.g., "foreign_key_error".
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:47]: Are we going to trim off `_error` too?

I think trimming this off for most notices would work ok (`InvalidRowLengthError`, `DuplicateKeyError`, `MissingRequiredFieldError`) - the only notices that wouldn't work for that I saw are:
* `ForeignKeyError` (maybe this should be `MissingPrimaryKeyError` or `OrphanedForeignKeyError` instead?)
* `IOError`
* `ThreadExecutionError`

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 1 above)
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:27]: I believe this should be `_notice`?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 3]
```diff
@@ -40,11 +42,18 @@ public SeverityLevel getSeverityLevel() {
...
    */
-  public abstract String getCode();
+  public String getCode() {
+    String toReturn =
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:50]: A more efficient implementation would be:

```java
String code = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, getClass().getSimpleName());
if (code.endsWith(NOTICE_SUBSTRING)) {
    return code.substring(0, code .length() - NOTICE_SUBSTRING.length());
}
return code;
```

...so the VM doesn't have to search for the string.

**Review by barbeau — COMMENTED** [review]
Thanks @lionel-nj! Some comments below.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 2 above)
Comment (lionel-nj) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:47]: Should these notice have their own implementation of `getCode`?

Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:47]: @barbeau No, we are not going to trim "error". Please see my plan that I have described earlier in https://github.com/MobilityData/gtfs-validator/pull/783#issuecomment-790123561

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 4]
```diff
@@ -40,11 +42,17 @@ public SeverityLevel getSeverityLevel() {
...
   /**
-   * Returns a descriptive type-specific name for this notice.
+   * Returns a descriptive type-specific name for this notice based on the class simple name.
    *
    * @return notice code, e.g., "foreign_key_error".
    */
-  public abstract String getCode();
+  public String getCode() {
+    String code = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, getClass().getSimpleName());
+    if (code.endsWith(NOTICE_SUFFIX)) {
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:51]: Please use org.apache.commons.lang3.StringUtils.removeEnd function.

I am encouraging you to google for ready library functions to perform basic operations in Java. That will make our code more idiomatic and readable.

**Review by aababilov — CHANGES_REQUESTED** [review]
I am on vacation this week. Please could you wait until I am back on 15 March?

**Comment by aababilov** [comment]
I am back! Looking forward to updates on this nice PR!

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 4 above)
Comment (lionel-nj) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java]: Changes have been provided in https://github.com/MobilityData/gtfs-validator/pull/788/commits/ddc44ccd9ee4ff238ddda1bfe907615bbe27b962

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 5]
```diff
@@ -40,11 +42,13 @@ public SeverityLevel getSeverityLevel() {
...
 
   /**
-   * Returns a descriptive type-specific name for this notice.
+   * Returns a descriptive type-specific name for this notice based on the class simple name.
    *
    * @return notice code, e.g., "foreign_key_error".
    */
-  public abstract String getCode();
+  public String getCode() {
+    return StringUtils.removeEnd(getClass().getSimpleName(), NOTICE_SUFFIX);
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:50]: Class name is UpperCamelCase, so it does not have "_notice" suffix. I think that you need to use CaseFormat. A unit test would be a great way to catch possible bugs here.

**Review by aababilov — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 5 above)
Comment (lionel-nj) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java]: Fixed in https://github.com/MobilityData/gtfs-validator/pull/788/commits/eecd38979e57f5bf5ad8d5db91356520aa975bc4.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 6]
```diff
@@ -40,11 +43,16 @@ public SeverityLevel getSeverityLevel() {
...
-   * Returns a descriptive type-specific name for this notice.
+   * Returns a descriptive type-specific name for this notice based on the class simple name.
    *
    * @return notice code, e.g., "foreign_key_error".
    */
-  public abstract String getCode();
+  public String getCode() {
+
+    return StringUtils.removeEnd(
+        CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, getClass().getSimpleName()),
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:53]: Class name is actually in UPPER_CAMEL (UpperCamel).

**Comment by aababilov** [comment]
Great job! Just one comment - and please submit the PR, I can't wait to see it in the repo.

**Comment by lionel-nj** [comment]
Thanks @aababilov!

**Pull request merged**